File: intrinsic/assets/instances/connect/connect.go
// Copyright 2026 Intrinsic Innovation LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package connect provides utility functions for connecting to gRPC services defined by GrpcConnection messages.
package connect

import (
	"context"
	"errors"
	"fmt"
	"strings"

	log "github.com/golang/glog"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/metadata"

	gcpb "intrinsic/assets/proto/v1/grpc_connection_go_proto"
)

// errNilConnection is returned when the provided GrpcConnection is nil.
var errNilConnection = errors.New("connection is nil")

// Connect creates a gRPC client connection for the specified GrpcConnection and
// returns a new context with attached metadata.
func Connect(ctx context.Context, connection *gcpb.GrpcConnection) (*grpc.ClientConn, context.Context, error) {
	if connection == nil {
		return nil, nil, errNilConnection
	}

	var metadataStrs []string
	for _, m := range connection.GetMetadata() {
		metadataStrs = append(metadataStrs, fmt.Sprintf("%s=%s", m.GetKey(), m.GetValue()))
	}
	log.Infof("Connecting to address %q with headers added to context: [%s]",
		connection.GetAddress(), strings.Join(metadataStrs, ", "))

	conn, err := grpc.NewClient(connection.GetAddress(), grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, nil, fmt.Errorf("failed to create gRPC client for address %q: %w", connection.GetAddress(), err)
	}

	// Add any needed metadata to the context.
	for _, m := range connection.GetMetadata() {
		ctx = metadata.AppendToOutgoingContext(ctx, m.GetKey(), m.GetValue())
	}

	return conn, ctx, nil
}

File: intrinsic/assets/instances/connect/connect.py
# Copyright 2026 Intrinsic Innovation LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Provides utility functions for connecting to gRPC services defined by GrpcConnection messages."""

from typing import Any
from typing import Sequence

from absl import logging
import grpc

from intrinsic.assets.proto.v1 import grpc_connection_pb2
from intrinsic.util.grpc import interceptor


def connect(
    connection: grpc_connection_pb2.GrpcConnection,
    grpc_options: Sequence[tuple[str, Any]] | None = None,
) -> grpc.Channel:
  """Creates a gRPC channel for communicating with the specified provider.

  The returned channel will be intercepted to include any needed metadata for
  communicating with the provider.

  Args:
    connection: The GrpcConnection message containing address and metadata.
    grpc_options: Additional options for the gRPC channel.

  Returns:
    A gRPC channel.
  """
  metadata = connection.metadata
  headers = [f"{m.key}={m.value}" for m in metadata]
  logging.info(
      'Connecting to address "%s" with headers injected through'
      " HeaderAdderInterceptor: [%s]",
      connection.address,
      ", ".join(headers),
  )

  channel = grpc.intercept_channel(
      grpc.insecure_channel(connection.address, grpc_options),
      interceptor.HeaderAdderInterceptor(
          lambda: [(m.key, m.value) for m in metadata]
      ),
  )

  return channel
